apollo-router/src/router.rs

```diff
@@ -464,121 +411,39 @@ pub(crate) enum Event {
-    /// This method can only be called once, and is not designed for use in dynamic configuration
-    /// scenarios.
-    ///
-    /// returns: Option<SocketAddr>
-    pub async fn ready(&mut self) -> Option<ListenAddr> {
-        self.state_receiver()
-            .map(|state| {
-                if let State::Running { address, .. } = state {
-                    Some(address)
-                } else {
-                    None
-                }
-            })
-            .filter(|socket| future::ready(socket != &None))
-            .map(|s| s.unwrap())
-            .next()
-            .boxed()
-            .await
-    }
-
-    /// Return a receiver of lifecycle events for the server. This method may only be called once.
-    ///
-    /// returns: mspc::Receiver<State>
-    pub fn state_receiver(&mut self) -> mpsc::Receiver<State> {
-        self.state_receiver.take().expect(
-            "State listener has already been taken. 'ready' or 'state' may be called once only.",
-        )
-    }
-
-    /// Trigger and wait until the server has shut down.
-    ///
-    /// returns: Result<(), FederatedServerError>
-    pub async fn shutdown(mut self) -> Result<(), FederatedServerError> {
-        self.maybe_close_state_receiver();
-        if self.shutdown_sender.send(()).is_err() {
-            tracing::error!("Failed to send shutdown event")
-        }
-        self.result.await
-    }
-
-    /// If the state receiver has not been set it must be closed otherwise it'll block the
-    /// state machine from progressing.
-    fn maybe_close_state_receiver(&mut self) {
-        if let Some(mut state_receiver) = self.state_receiver.take() {
-            state_receiver.close();
+impl RouterHandle {
+    /// Returns the server handle (self) when the router is ready to receive requests.
+    /// Note that the router may not start
+    pub async fn ready(self) -> Result<Self, FederatedServerError> {
+        if *self.ready.read().await {
+            Ok(self)
+        } else {
+            Err(FederatedServerError::StartupError)
         }
     }
+}
 
-    /// State receiver that prints out basic lifecycle events.
-    pub async fn with_default_state_receiver(&mut self) {
-        self.state_receiver()
-            .for_each(|state| {
-                match state {
-                    State::Startup => {
-                        tracing::info!("starting Apollo Router")
-                    }
-                    State::Running { .. } => {}
-                    State::Stopped => {
-                        tracing::info!("stopped")
-                    }
-                    State::Errored => {
-                        tracing::info!("stopped with error")
-                    }
-                }
-                future::ready(())
-            })
-            .await
+impl Drop for RouterHandle {
+    fn drop(&mut self) {
+        let _ = self
+            .shutdown_sender
+            .take()
+            .expect("shutdown sender mut be present")
```

Comment:
```suggestion
            .expect("shutdown sender must be present")
```

Reply:
Fixed in [992ddd6](https://github.com/apollographql/router/pull/1227/commits/992ddd6ae411c6fdb4b9a5d39a21132ff8a00c48)